tests/basis_test.py

```diff
@@ -1,330 +1,438 @@
-    fn = ChebyshevSecondKind(5)
-    assert numpy.allclose(f0(xs),special.eval_chebyu(0,xs))
-    assert numpy.allclose(f1(xs),special.eval_chebyu(1,xs))
-    assert numpy.allclose(fn(xs),special.eval_chebyu(5,xs))
-
-
-def test_cheb_2nd_call_invalid_input():
-    """Test call raises error if input is outside domain [-1,1]"""
-    f = ChebyshevSecondKind(4)
+        basis_fun(too_large)
     with pytest.raises(ValueError):
-        f(2)
+        basis_fun(too_small)
     with pytest.raises(ValueError):
-        f(-2)
+        basis_fun([valid_input, too_small, valid_input])
     with pytest.raises(ValueError):
-        f([0.5,0.7,3,0.8])
+        basis_fun(
+            [
+                [valid_input, too_large, valid_input],
+                [valid_input, valid_input, valid_input],
+            ]
+        )
 
 
-def test_cheb_2nd_derivative_invalid_input():
-    """Test call raises error if input is outside domain [-1,1]"""
-    f = ChebyshevSecondKind(4)
+@basis_outside_domain
+def test_derivative_outside_domain_error(basis_fun, too_large, too_small, valid_input):
+    """Test derivative raises error for input outside domain"""
     with pytest.raises(ValueError):
-        f.derivative(2)
+        basis_fun.derivative(too_large)
     with pytest.raises(ValueError):
-        f.derivative(-2)
+        basis_fun.derivative(too_small)
     with pytest.raises(ValueError):
-        f.derivative([0.5,0.7,3,0.8])
-
-
-def test_set_initialize_empty():
-    """Check BasisFunctionSet initializes with empty set"""
-    basis_functions = []
-    points = []
-    f = BasisFunctionSet(points,basis_functions)
-    assert f.points == []
-    assert f.basis_functions == []
-
-def test_set_initialize_0():
-    """Check NestedBasisFunctionSet correctly initializes"""
-    basis_functions = [ChebyshevFirstKind(0)]
-    points = [0]
-    f = BasisFunctionSet(points,basis_functions)
-    assert f.points == [0]
-    assert f.basis_functions[0].n == 0
-
-def test_set_invalid_input():
-    """Check BasisFunctionSet gives error for invalid inputs"""
-    basis_functions = [ChebyshevFirstKind(0)]
-    points = [0,1,2]
-    with pytest.raises(IndexError):
-        f = BasisFunctionSet(points,basis_functions)
-
-def test_set_nested_initialize_empty():
-    """Check NestedBasisFunctionSet initializes with empty set"""
-    levels = []
-    basis_functions = []
-    points = []
-    f = NestedBasisFunctionSet(points,basis_functions,levels)
-    assert f.points == []
-    assert f.levels == []
+        basis_fun.derivative([valid_input, too_small, valid_input])
+    with pytest.raises(ValueError):
+        basis_fun.derivative(
+            [
+                [valid_input, too_large, valid_input],
+                [valid_input, valid_input, valid_input],
+            ]
+        )
+
+
+# Test call function correctness
+@basis_call_answer_key
+def test_call(basis_fun, answer_key):
+    """Test basis function call"""
+    print(basis_fun)
+    print(answer_key)
```

Comment:
Remove print statements from tests. If there is an issue where the parametrized tests are not showing up with nice names, there is a way to get nicer names with pytest that we can look at!

Reply:
I've added ids to these tests and the benchmark function tests.